### PR TITLE
fix(hir): propagate ws Client handle class across function-call boundaries

### PR DESCRIPTION
**Pull request by machineloop**
## Summary

The `("ws","Client")` handle delivered to a WebSocket-upgrade callback
`server.on("upgrade", (req, wsId, head) => …)` (raw `node:http`) only dispatched its
`.send()`/`.on()`/`.close()` to the dedicated `js_ws_*_client_i64` runtime when used **inline**
in the callback. Passing `wsId` into a helper made those calls a **silent no-op** — the frame
was dropped with no error. This propagates the handle's host class across the call boundary so
the helper dispatches correctly.

## Changes

- `crates/perry-hir/src/lower/pre_scan.rs` — new `pre_scan_cross_fn_native_params` pass (run
  before any function body lowers). It seeds from the upgrade idiom `X.on("upgrade", (req,
  wsId, head) => …)` — **gated on `X` being a `createServer(...)` result** (static proxy for a
  `("http","HttpServer")` instance; mirrors the runtime check, avoids false positives on
  arbitrary `.on("upgrade")`) — and follows the handle through `userFn(…, wsId, …)` calls
  **transitively** (fixpoint), recording a `(fn_name, param_index) -> (module, class)` hint.
- `crates/perry-hir/src/lower/lowering_context.rs` + `…/lower/context.rs` — add the
  `param_native_hints` map.
- `crates/perry-hir/src/lower/lower_module_fn.rs` — invoke the pass alongside the other pre-scans.
- `crates/perry-hir/src/lower_decl/fn_decl.rs` — apply the hint (only when the param isn't
  already a native instance, so an explicit annotation wins) before the body lowers.
- `crates/perry/tests/ws_client_handle_cross_function_dispatch.rs` — regression tests.

## Test plan

```
cargo test -p perry --test ws_client_handle_cross_function_dispatch   # 4 passed
cargo fmt -p perry-hir -p perry --check                               # clean
cargo build --release --workspace --exclude perry-ui-{ios,tvos,watchos,gtk4,android,windows}
```

A live WS round-trip needs a socket + client and Perry's raw-upgrade `wsId` has no
Node-`ws`-compatible API to diff against, so the bug (which runtime symbol the call site lowers
to) is asserted at the LLVM-IR level. TDD red→green: for the helper-only case the IR went from
**0** `call @js_ws_send_client_i64` sites (silent no-op) to **1**. Tests also cover transitive
forwarding, `.on`/`.close`, and a control proving a non-ws `.send` param keeps running its own
method (no mis-tagging).

- [x] `cargo build --release -p perry` clean (the crates this change touches). The full
      `--workspace` release build fails only on `windows-future` — a Windows-only transitive
      dep that can't compile on this macOS host (env limitation, unrelated to this change).
- [ ] `cargo test --workspace …` passes — my new tests pass; the suite's only failures
      (`integer_locals_provenance`, `functional_batch2_regressions`, and a
      `blocklist_addsubnet_prefix` `TempDir::join` compile break) reproduce **identically on
      clean `main` without this change**, so they are pre-existing and unrelated.
- [x] (if user-facing) Added a `#[test]` in the affected crate (`crates/perry/tests/`)
- [ ] (if CLI / stdlib / runtime API changed) Updated `docs/src/` — N/A (internal HIR dispatch)
- [ ] (if touching a platform UI backend) — N/A

## Screenshots / output

```
function handleConnection(req, wsId) { wsId.send("hi"); }   // before: dropped; after: delivered
server.on("upgrade", (req, wsId, head) => { handleConnection(req, wsId); });
// IR: helper `wsId.send` -> `call void @js_ws_send_client_i64(...)`  (0 sites before, 1 after)
```

## Checklist

- [x] I have NOT bumped the workspace version or edited CLAUDE.md / CHANGELOG.md
- [x] My commits follow the loose `fix:` prefix convention
- [x] I've read CONTRIBUTING.md and agree to the Code of Conduct

## Summary by CodeRabbit

* **Bug Fixes / New Features**
  * Improved WebSocket upgrade-handle behavior so `send`, `on`, and `close` dispatch to the correct Client runtime even when the `wsId` handle is forwarded through helper call chains, including nested callbacks.
  * Added safeguards for edge cases such as parameter shadowing, polymorphic forwarding, and correct behavior when mixed with non-WebSocket objects.
* **Tests**
  * Expanded the WebSocket cross-function dispatch regression suite with helper chains, negative controls, and additional scenarios for parameter indexing and `createServer` provenance.

### CODE REVIEW

**Comment by machineloop** [comment]
> **Actionable comments posted: 4**
@coderabbitai autofix stacked pr

File: crates/perry-hir/src/lower/pre_scan.rs [hunk 1]
```diff
@@ -266,6 +266,604 @@ pub(crate) fn pre_scan_mixin_functions(ast_module: &ast::Module, ctx: &mut Lower
...
+        let Some(body) = &fd.function.body else {
+            continue;
+        };
+        let name = fd.ident.sym.to_string();
+        let names: Vec<Option<String>> = fd
+            .function
+            .params
+            .iter()
+            .map(|p| cross_fn_pat_name(&p.pat))
+            .collect();
```
[the remark this answers is not in the document]
Comment (machineloop) [reply crates/perry-hir/src/lower/pre_scan.rs:323]: Fixed in e84da080 — the pre-scan now drops the TypeScript `this:` param when building its param list, matching `lower_fn_decl` (which skips `this` before enumerating real params). Hint/param indices stay aligned for `function f(this: T, req, wsId)`. Regression test: `leading_this_param_keeps_hint_index_aligned`.

File: crates/perry-hir/src/lower/pre_scan.rs [hunk 2]
```diff
@@ -266,6 +266,604 @@ pub(crate) fn pre_scan_mixin_functions(ast_module: &ast::Module, ctx: &mut Lower
...
+        };
+        let name = match callee.as_ref() {
+            ast::Expr::Ident(i) => i.sym.to_string(),
+            ast::Expr::Member(m) => match &m.prop {
+                ast::MemberProp::Ident(i) => i.sym.to_string(),
+                _ => return false,
+            },
+            _ => return false,
+        };
+        name == "createServer" || name == "createSecureServer"
```
[the remark this answers is not in the document]
Comment (machineloop) [reply crates/perry-hir/src/lower/pre_scan.rs]: Fixed in 1b58b78c — `createServer`/`createSecureServer` is now only treated as a server seed when it resolves to a verified `node:http`/`node:https`/`node:http2` import or `require(...)` alias (named import, namespace member, or destructured require), so a user-defined factory or unrelated `obj.createServer()` no longer seeds `("ws","Client")` taint. Regression test: `non_http_create_server_does_not_seed_ws_taint`.

File: crates/perry-hir/src/lower/lowering_context.rs [hunk 3]
```diff
@@ -193,6 +193,15 @@ pub struct LoweringContext {
...
     pub(crate) native_instances: Vec<(String, String, String)>,
+    /// Cross-function native-instance hints: (function_name, param_index) ->
+    /// (module_name, class_name). Populated by `pre_scan_cross_fn_native_params`
+    /// BEFORE any function body is lowered, so `lower_fn_decl` can tag an
+    /// otherwise-untyped parameter as a native instance when a known native
+    /// handle (e.g. the `("ws","Client")` upgrade `wsId`) is passed across the
+    /// call boundary into it. Without this, `wsId.send(...)` inside a helper
+    /// reached as `handleConnection(req, wsId)` lowers to a generic (silent
+    /// no-op) dispatch instead of `js_ws_send_client_i64`.
+    pub(crate) param_native_hints: HashMap<(String, usize), (String, String)>,
```
[the remark this answers is not in the document]
Comment (machineloop) [reply crates/perry-hir/src/lower/lowering_context.rs]: Fixed in 319c88da — `param_native_hints` is now keyed by the function declaration's identifier span (a stable AST identity — its `lo` byte offset) instead of its name, at both the pre-scan record site and the `lower_fn_decl` apply site. A hint seeded for one declaration can no longer leak onto a same-named nested/redeclared function. Regression test: `same_named_function_is_not_cross_tagged` (asserts exactly one Client dispatch in the module).

File: crates/perry-hir/src/lower/pre_scan.rs [hunk 4]
```diff
@@ -266,6 +266,604 @@ pub(crate) fn pre_scan_mixin_functions(ast_module: &ast::Module, ctx: &mut Lower
...
+            ast::BlockStmtOrExpr::Expr(e) => collect_calls_in_expr(e, out),
+        },
+        ast::Expr::Fn(f) => {
+            if let Some(b) = &f.function.body {
+                collect_calls_in_stmts(&b.stmts, out);
+            }
+        }
+        _ => {}
+    }
+}
```
[the remark this answers is not in the document]
Comment (machineloop) [reply crates/perry-hir/src/lower/pre_scan.rs]: Fixed in cbf05ecd — propagation is now a scope-aware walk (`walk_taint_*`) that carries a per-scope taint set and drops any name a nested function/arrow rebinds as a parameter before recursing. A shadowing `function later(wsId) {…}` no longer inherits the outer Client handle, while legitimate captures (`wsId.on("message", m => helper(req, wsId))`) still propagate. Regression test: `shadowing_nested_param_does_not_inherit_taint`.

**Comment by machineloop** [comment]
Addressed the remaining nitpick (`Expr::Class` / `TsSatisfies` not traversed in `walk_taint_expr`) in d67a1bed: class method/constructor bodies are now traversed with their params pruned (like object methods), plus field initializers and static blocks, and `TsSatisfies` is unwrapped like the other TS assertion wrappers. Rebased onto the latest branch head. All 11 ws-dispatch tests pass.
